src/types/program.rs

```diff
@@ -32,10 +32,17 @@ impl Program {
     }
 
     pub fn from_reader(
-        reader: impl Read,
+        reader: impl Read + Copy,
         entrypoint: Option<&str>,
     ) -> Result<Program, ProgramError> {
-        deserialize_program(reader, entrypoint)
+        match deserialize_program(reader, entrypoint) {
+            Ok(program) => return Ok(program),
+            Err(e) => {
+                println!("{}", e);
+                println!("Failed to deserialize the program using the old cairo-lang syntax, falling back to the new syntax");
+            }
+        }
+        deserialize_program_new_syntax(reader, entrypoint)
```

Comment:
Please let me know whether this is okay. The goal is to fall back to the new syntax automatically and to keep the error from the old syntax deserialization at the same time.

I'm not sure about the output, but it feels like a good idea to let users know that one deserialization has failed, so I made it explicit.